File: c++/array/mergeSort.cpp
class Solution {
public:
    vector<int> merge(vector<int> left, vector<int> right) {
        vector<int> temp;
        int i = 0;
        int j = 0;

        while (i < left.size() && j < right.size()) {
            if (left[i] < right[j])
                temp.push_back(left[i++]);
            else {
                temp.push_back(right[j++]);
            }
        }
        while(i<left.size()){
            temp.push_back(left[i++]);
        }
        while(j<right.size()){
            temp.push_back(right[j++]);
        }

        return temp;
    }

    vector<int> sortArray(vector<int>& nums) {
        if (nums.size() <= 1)
            return nums;
        int mid = nums.size() / 2;

        vector<int> left(nums.begin(), nums.begin() + mid);
        vector<int> right(nums.begin() + mid, nums.end());

        return merge(sortArray(left), sortArray(right));
    }
};

File: javascript/mergeSort.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
function merge(left, right) {
    let res = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            res.push(left[i++]);
        } else {
            res.push(right[j++]);
        }
    }

    return [...res, ...left.slice(i), ...right.slice(j)];
}

var sortArray = function (nums) {
    if (nums.length <= 1) return nums;
    let mid = Math.floor(nums.length / 2);
    let left = nums.slice(0, mid);
    let right = nums.slice(mid);

    return merge(sortArray(left), sortArray(right));
};